// convex/dist/esm/server/impl/search_filter_builder_impl.js
"use strict";
import { convexToJson } from "../../values/values.js";
import {
  SearchFilter
} from "../search_filter_builder.js";
export class SearchFilterBuilderImpl extends SearchFilter {
  constructor(filters) {
    super();
    this.filters = filters;
    this.isConsumed = false;
  }
  static new() {
    return new SearchFilterBuilderImpl([]);
  }
  consume() {
    if (this.isConsumed) {
      throw new Error(
        "SearchFilterBuilder has already been used! Chain your method calls like `q => q.search(...).eq(...)`."
      );
    }
    this.isConsumed = true;
  }
  search(fieldName, query) {
    this.consume();
    return new SearchFilterBuilderImpl(
      this.filters.concat({
        type: "Search",
        fieldPath: fieldName,
        value: query
      })
    );
  }
  eq(fieldName, value) {
    this.consume();
    return new SearchFilterBuilderImpl(
      this.filters.concat({
        type: "Eq",
        fieldPath: fieldName,
        value: convexToJson(value)
      })
    );
  }
  export() {
    this.consume();
    return this.filters;
  }
}
