// convex/src/server/impl/actions_impl.ts
import { convexToJson, jsonToConvex } from "../../values/index.js";
import { version } from "../../index.js";
import { performAsyncSyscall } from "./syscall.js";

export interface ConvexActionDatabaseClient {
  runQuery(name: string, ...args: any[]): Promise<any>;
  runMutation(name: string, ...args: any[]): Promise<any>;
}

export function setupActionDatabaseClient(
  requestId: string
): ConvexActionDatabaseClient {
  return {
    runQuery: async (name, ...args) => {
      const syscallArgs = {
        name,
        args: convexToJson(args),
        version,
        requestId,
      };
      const result = await performAsyncSyscall("actions/query", syscallArgs);
      return jsonToConvex(result);
    },
    runMutation: async (name, ...args) => {
      const syscallArgs = {
        name,
        args: convexToJson(args),
        version,
        requestId,
      };
      const result = await performAsyncSyscall("actions/mutation", syscallArgs);
      return jsonToConvex(result);
    },
  };
}
