// convex/dist/cjs/server/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var server_exports = {};
__export(server_exports, {
  HttpRouter: () => import_router.HttpRouter,
  actionGeneric: () => import_registration_impl.actionGeneric,
  httpEndpointGeneric: () => import_registration_impl.httpEndpointGeneric,
  httpRouter: () => import_router.httpRouter,
  mutationGeneric: () => import_registration_impl.mutationGeneric,
  queryGeneric: () => import_registration_impl.queryGeneric
});
module.exports = __toCommonJS(server_exports);
__reExport(server_exports, require("./database.js"), module.exports);
__reExport(server_exports, require("./data_model.js"), module.exports);
var import_registration_impl = require("./impl/registration_impl.js");
__reExport(server_exports, require("./pagination.js"), module.exports);
__reExport(server_exports, require("./search_filter_builder.js"), module.exports);
__reExport(server_exports, require("./storage.js"), module.exports);
__reExport(server_exports, require("./scheduler.js"), module.exports);
var import_router = require("./router.js");
